File: node-service/internal/application/usecases/listProducts.js
const { toResponseList } = require('../dto/productDto');

class ListProductsUseCase {
  constructor(repository, logger, cacheService) {
    this.repository = repository;
    this.logger = logger;
    this.cache = cacheService;
  }

  async execute() {
    const loader = async () => {
      const products = await this.repository.list();
      return toResponseList(products);
    };

    const { value, status } = await this.cache.fetch('products:list', loader);
    this.annotateCacheStatus(value, status);
    this.logger.info('listed products', { count: value.length, cacheStatus: status });
    return value;
  }

  annotateCacheStatus(products, status) {
    if (!status || !Array.isArray(products)) {
      return;
    }
    for (const product of products) {
      product.cache_status = status;
    }
  }
}

module.exports = ListProductsUseCase;

File: go-service/internal/application/usecase/list_products.go
package usecase

import (
	"context"

	"github.com/example/interview-stack/go-service/internal/application/cache"
	"github.com/example/interview-stack/go-service/internal/application/dto"
	"github.com/example/interview-stack/go-service/internal/domain"
)

// ListProductsUseCase handles retrieval of all products.
type ListProductsUseCase struct {
	repo   domain.ProductRepository
	logger domain.Logger
	cache  *cache.Service
}

func NewListProductsUseCase(repo domain.ProductRepository, logger domain.Logger, cacheSvc *cache.Service) *ListProductsUseCase {
	return &ListProductsUseCase{repo: repo, logger: logger, cache: cacheSvc}
}

func (uc *ListProductsUseCase) Execute(ctx context.Context) ([]dto.ProductResponse, error) {
	loader := func(ctx context.Context) ([]dto.ProductResponse, error) {
		products, err := uc.repo.ListProducts(ctx)
		if err != nil {
			return nil, err
		}
		return dto.FromDomainProducts(products), nil
	}

	result, meta, err := cache.Fetch(uc.cache, ctx, "products:list", loader)
	if err != nil {
		return nil, err
	}
	annotateCacheStatus(result, meta.Status)
	uc.logger.Info("list products executed", domain.KV("count", len(result)), domain.KV("cache_status", meta.Status))
	return result, nil
}

func annotateCacheStatus(products []dto.ProductResponse, status cache.Status) {
	if status == "" {
		return
	}
	for i := range products {
		products[i].CacheStatus = string(status)
	}
}
